test/jdk/java/nio/channels/Selector/KeyCancelled.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/* @test
+ * @bug 8287580
+ * @summary Check CancelledKeyException not thrown during channel registration
+ * @run main KeyCancelled
+ */
+
+import java.io.IOException;
+import java.nio.channels.SelectionKey;
+import java.nio.channels.Selector;
+import java.nio.channels.SocketChannel;
+import java.util.concurrent.CountDownLatch;
+
+public class KeyCancelled {
+    private static final int ITERATIONS = 4;
+
+    public static void main(String[] args) throws Exception {
+        for (int iteration = 1; iteration <= 4; iteration++) {
+            System.out.printf("Iteration %d%n", iteration);
+            test();
+        }
+    }
+
+    private static void test() throws IOException {
+        Thread t;
+        try (Selector s = Selector.open()) {
+            // Use latch to avoid a ClosedSelectorException
+            final CountDownLatch latch = new CountDownLatch(1);
+            t = new Thread(() -> {
+                for (int i = 0; i < 100_000; i++) {
+                    s.keys().forEach(SelectionKey::cancel);
+                }
+                latch.countDown();
+            });
+            t.start();
+
+            for (int i = 0; i < 10_000; i++) {
+                try (SocketChannel c = s.provider().openSocketChannel()) {
+                    c.configureBlocking(false);
+                    // Sometimes this throws CancelledKeyException, because
+                    // the key is cancelled by the other thread part-way through
+                    // the register call.
+                    c.register(s, SelectionKey.OP_READ);
+                    // c.isRegistered() is false here after the exceptional case
+                }
```

Comment:
This tries to create 10k SocketChannels. I would be concerned with the robustness of this test. I suspect it should be possible to create the conditions for the bug with just one socket channel where the registration key is flushed from the Selector in the loop, the next registration will create a new key that can be cancelled asynchronously by the other thread.

Reply:
Commit 925336449bfda90791b7e289d0008da4a9bc87a6 changes the test to use a single `SocketChannel`.

Reply:
Commit 98bcd22e8ba60c7499fab332f943616bb88b03f5 reduces the number of `SocketChannels` created per invocation of `test()` to 5. With this change, a run of 5 test repeats in the test harness had these failure rates:
```
5/5 failures linux-aarch64
4/5 failures linux-x64
4/5 failures macosx-x64
3/4 failures windows-x64
```
This looks sufficient to detect a failure but with a much lower number of open sockets.

Reply:
When i ran you test code(without the fix) on my mac , i am getting sometime NPE as below. Please have look, your test may fail intermittently in future with below NPE.

/Library/Java/JavaVirtualMachines/jdk1.8.0_321.jdk/Contents/Home/bin/java 
i: 0
i: 1
Exception in thread "Thread-0" java.nio.channels.ClosedSelectorException
	at sun.nio.ch.SelectorImpl.keys(SelectorImpl.java:68)
	at com.vmware.test.KeyCancelled.lambda$test$0(KeyCancelled.java:25)
	at java.lang.Thread.run(Thread.java:750)
Exception in thread "main" java.nio.channels.CancelledKeyException
	at sun.nio.ch.SelectionKeyImpl.ensureValid(SelectionKeyImpl.java:73)
	at sun.nio.ch.SelectionKeyImpl.interestOps(SelectionKeyImpl.java:82)
	at sun.nio.ch.SelectorImpl.register(SelectorImpl.java:134)
	at java.nio.channels.spi.AbstractSelectableChannel.register(AbstractSelectableChannel.java:212)
	at java.nio.channels.SelectableChannel.register(SelectableChannel.java:280)
	at com.vmware.test.KeyCancelled.test(KeyCancelled.java:38)
	at com.vmware.test.KeyCancelled.main(KeyCancelled.java:14)
	Suppressed: java.lang.NullPointerException
		at java.nio.channels.spi.AbstractSelectableChannel.removeKey(AbstractSelectableChannel.java:131)
		at java.nio.channels.spi.AbstractSelector.deregister(AbstractSelector.java:188)
		at sun.nio.ch.KQueueSelectorImpl.implClose(KQueueSelectorImpl.java:209)
		at sun.nio.ch.SelectorImpl.implCloseSelector(SelectorImpl.java:113)
		at java.nio.channels.spi.AbstractSelector.close(AbstractSelector.java:111)
		at com.vmware.test.KeyCancelled.test(KeyCancelled.java:47)
		... 1 more